File: tests/typesafety/python/carrier_protocol_misuse.py
# expect: fails
# title: A carrier whose quote returns text, rejected by the checker
# misuse-of: carrier_protocol
from typing import Protocol


class Carrier(Protocol):
    def quote_pence(self, kilograms: float) -> int: ...


class ChattyCourier:
    def quote_pence(self, kilograms: float) -> str:
        return f"about {int(kilograms) * 100} pence"


def cheapest(carriers: list[Carrier], kilograms: float) -> int:
    return min(carrier.quote_pence(kilograms) for carrier in carriers)


best = cheapest([ChattyCourier()], 9.5)

File: tests/typesafety/typescript/carrier_protocol_misuse.ts
// expect: fails

type Carrier = {
  quotePence(kilograms: number): number;
};

class ChattyCourier {
  quotePence(kilograms: number): string {
    return `about ${Math.floor(kilograms) * 100} pence`;
  }
}

function cheapest(carriers: readonly Carrier[], kilograms: number): number {
  return Math.min(...carriers.map((carrier) => carrier.quotePence(kilograms)));
}

export const best = cheapest([new ChattyCourier()], 9.5);
